libwnbd/wnbd_ioctl.cpp

```diff
@@ -139,6 +144,367 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
+    }
+
+    for (DWORD I = 0; !Found && SetupDiEnumDeviceInfo(*AdapterDevHandle, I, DevInfoData); I++)
+    {
+        if (!SetupDiGetDeviceRegistryPropertyA(*AdapterDevHandle, DevInfoData, SPDRP_HARDWAREID, 0,
+            (PBYTE)TempBuf, sizeof(TempBuf) - 1, 0)) {
+            continue;
+        }
+
+        if (strstr(TempBuf, WNBD_HARDWAREID) != NULL) {
+            Found = TRUE;
+            break;
+        }
+    }
+
+    if (!Found) {
+        Status = GetLastError();
+        if (ERROR_NO_MORE_ITEMS != Status) {
+            LogError(
+                "Failed to locate device with hardware id: %s. Error: %d. Error message: %s",
+                WNBD_HARDWAREID, Status, win32_strerror(Status).c_str());
+        }
+        return Status;
+    }
+
+    return ERROR_SUCCESS;
+}
+
+DWORD RemoveAllDevices(BOOLEAN Force)
+{
+    DWORD BufferSize = 0;
+    DWORD Status = 0;
+    PWNBD_CONNECTION_LIST ConnList = NULL;
+
+    WNBD_OPTION_VALUE OptValue = { WnbdOptBool };
+    /* Disallow new mappings so we can remove all current mappings */
+    OptValue.Data.AsBool = FALSE;
+    Status = WnbdSetDrvOpt("NewMappingsAllowed", &OptValue, FALSE);
+    if (Status) {
+        goto exit;
+    }
+
+    Status = WnbdList(ConnList, &BufferSize);
+    if (!BufferSize) {
+        goto exit;
+    }
+    ConnList = (PWNBD_CONNECTION_LIST)calloc(1, BufferSize);
+    if (NULL == ConnList) {
+        Status = ERROR_NOT_ENOUGH_MEMORY;
+        LogError(
+            "Failed to allocate %d bytes.", BufferSize);
+        goto exit;
+    }
+    Status = WnbdList(ConnList, &BufferSize);
+    if (Status) {
+        goto exit;
+    }
+    CHAR* InstanceName;
+    for (ULONG index = 0; index < ConnList->Count; index++) {
+        InstanceName = ConnList->Connections[index].Properties.InstanceName;
+        WNBD_REMOVE_OPTIONS RemoveOptions = { 0 };
+        RemoveOptions.Flags.HardRemoveFallback = TRUE;
+        RemoveOptions.SoftRemoveTimeoutMs = 30 * 1000;
+        RemoveOptions.SoftRemoveRetryIntervalMs = 1 * 1000;
+        RemoveOptions.Flags.HardRemove = Force;
+        Status = WnbdRemoveEx(InstanceName, &RemoveOptions);
+        if (Status) {
+            goto exit;
+        }
+    }
+
+exit:
+    if (ConnList) {
+        free(ConnList);
+    }
+    return Status;
+}
+
+DWORD WnbdUninstallDriver(PBOOL RebootRequired, BOOLEAN Force)
+{
+    DWORD Status = ERROR_SUCCESS;
+    HDEVINFO AdapterDevHandle = INVALID_HANDLE_VALUE;
+    SP_DEVINFO_DATA DevInfoData;
+    DevInfoData.cbSize = sizeof DevInfoData;
+
+    PWNBD_CONNECTION_LIST ConnList = NULL;
+    Status = RemoveAllDevices(Force);
+    if (Status) {
+        goto exit;
+    }
+
+    while (Status == ERROR_SUCCESS) {
+        Status = FindWnbdAdapterDevice(&AdapterDevHandle, &DevInfoData);
+        if (ERROR_SUCCESS != Status) {
+            goto exit;
+        }
+
+        Status = RemoveWnbdDevice(AdapterDevHandle, &DevInfoData, RebootRequired);
+
+        if (INVALID_HANDLE_VALUE != AdapterDevHandle) {
```

Comment:
Do we need this check? Wouldn't FindWnbdAdapterDevice return an error?

Reply:
Calling SetupDiDestroyDeviceInfoList on `INVALID_HANDLE_VALUE` is undefined AFAIR, but the documentation doesn't mention it so will remove it.